system/metrics/sinks/prometheus/prometheusSink.cpp

```diff
@@ -109,43 +113,141 @@ void PrometheusMetricSink::toPrometheusMetrics(const std::vector<std::shared_ptr
+        }
+        else
         {
-            out.append(" {");
-            bool firstEntry = true;
-            for (auto &metaDataIt: metaData)
+            out.append(metricName.c_str());
+            const auto &metaData = pMetric->queryMetaData();
+            if (!metaData.empty())
             {
-                if (!firstEntry)
-                    out.append(",");
-                else
-                    firstEntry=false;
+                out.append(" {");
+                bool firstEntry = true;
+                for (auto &metaDataIt: metaData)
+                {
+                    if (!firstEntry)
+                        out.append(",");
+                    else
+                        firstEntry = false;
 
-                out.append(metaDataIt.key.c_str()).append("=\"").append(metaDataIt.value.c_str()).append("\"");
+                    out.append(metaDataIt.key.c_str()).append("=\"").append(metaDataIt.value.c_str()).append("\"");
+                }
+                out.append("}");
             }
-            out.append("}");
+            out.append(" ").append(pMetric->queryValue()).append("\n");
         }
+    }
+}
+
+void PrometheusMetricSink::toPrometheusHistogram(const std::string &name, const std::shared_ptr<IMetric> &pHistogram, StringBuffer & out)
+{
+    auto labels = getHistogramLabels(name, pHistogram);
+    auto values = pHistogram->queryHistogramValues();
+    int index = 0;
+    __uint64 cumulativeCount = 0;
+    for (auto const &value: values)
+    {
+        cumulativeCount += value;
+        out.append(labels[index].c_str()).append(" ").append(cumulativeCount).append("\n");
+        ++index;
+    }
+
+    //
+    // The next two labels using indexes are the sum and count labels.
+    out.append(labels[index++].c_str()).append(" ").append(pHistogram->queryValue()).append("\n");
+    out.append(labels[index].c_str()).append(" ").append(cumulativeCount).append("\n");
+}
+
+
+std::string PrometheusMetricSink::getPrometheusMetricName(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::string name = pMetric->queryName();
+
+    //'.' is a known char used in HPCC metric names but invalid in Prometheus
+    std::replace(name.begin(), name.end(), '.', '_');
+    name.append("_").append(getPrometheusMetricUnits(pMetric));
+    return name;
+}
 
-        out.append(" ").append(pMetric->queryValue()).append("\n");
+std::string PrometheusMetricSink::getPrometheusMetricUnits(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::string unitsStr;
+    switch (pMetric->queryUnits())
+    {
+        case SMeasureCount:   unitsStr = "total";    break;
+        case SMeasureSize:    unitsStr = "bytes";    break;
+        case SMeasureTimeNs:  unitsStr = "seconds";  break;
+        default:              unitsStr = "";         break;
     }
+    return unitsStr;
 }
 
+const std::vector<std::string> &PrometheusMetricSink::getHistogramLabels(const std::string &name, const std::shared_ptr<IMetric> &pHistogram)
```

Comment:
this function should have self documenting block comment which describe the labels format and what they mean, for instance, what is ""_bucket{\"le=+Inf\""

Reply:
Added a block above the function